src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,177 @@
+                {
+                    "-1" => Channel.CreateUnbounded<string>(),              // buffer size of -1 indicates that buffer should be enabled but unbounded
+                    "0" => null,                                            // buffer size of 0 indicates that buffer should be disabled
+                    var s when int.TryParse(s, out int i) && i > 0 => Channel.CreateBounded<string>(i),
+                    _ => Channel.CreateBounded<string>(DefaultBufferSize),  // default behavior is to use buffer with default size
+                };
+
+                if (_consoleBuffer == null)
+                {
+                    _writeEvent = Console.WriteLine;
+                }
+                else
+                {
+                    _writeEvent = WriteToConsoleBuffer;
+                    _consoleBufferTimeout = consoleBufferTimeout;
+                    if (autoStart)
+                    {
+                        bool batched = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferBatched) switch
+                        {
+                            "0" => false,     // disable batching by setting to 0
+                            _ => true,        // default behavior is batched
+                        };
+
+                        _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+                    }
+                }
+            }
+            else
+            {
+                _writeEvent = (string s) => { };
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public void WriteHandler(string evt)
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
+                {
+                    // Buffer is currently full, wait until writing is permitted.
+                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
+                    if (writeTask.WaitAsync(_consoleBufferTimeout).Result == false)
+                    {
+                        // The buffer has been completed and does not allow further writes.
+                        // Currently this should not be possible, but the safest thing to do is just write directly to the console.
+                        Console.WriteLine(evt);
+                        break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
+                // dump the raw exception and write the event to the console directly
```

Comment:
Potentially under load, we might get into this timeout situation often. Won't this mean that logs will be out of order with out of sync timestamps? E.g. the forced Console write jumps the queue and is logged before other pending logs?

Reply:
Yes, this implementation can result in out of order logs. The alternative is to fully block the writes. Do you have a strong preference?

Reply:
Yeah, I'd say we should avoid out of order logs. If we're putting buffer in place for output I think we should honor the ordering, and block until we can write.

Reply:
As we discussed yesterday, it seems better to have out of order logs than to be blocking the execution code path on waiting for the buffer to have space, which is reliant on the background task that is processing the buffer. This only occurs when the buffer is full and when we've waited for the buffer to be cleared (currently 1 second). We have the EventTimestamp column that can be used in cases where ordering is critical.